evm/src/trace/identifier/signatures.rs

```diff
@@ -0,0 +1,135 @@
+use ethers::abi::{Event, Function};
+use foundry_utils::{decode_selector, get_event, get_func, selectors::SelectorType};
+use serde::{Deserialize, Serialize};
+use std::{collections::BTreeMap, io::BufWriter, path::PathBuf};
+use tracing::warn;
+
+/// As identifier that tries to identify functions and events using their signatures.
+#[derive(Debug, Default)]
+pub struct SignaturesIdentifier {
+    cached: CachedSignatures,
+    cached_path: Option<PathBuf>,
+}
+
+impl SignaturesIdentifier {
+    pub fn new(cache_path: Option<PathBuf>) -> eyre::Result<Self> {
+        if let Some(cache_path) = cache_path {
+            let path = cache_path.join("signatures");
+            let cached = if path.is_file() {
+                serde_json::from_reader(std::fs::File::open(&path)?)?
+            } else {
+                if let Err(err) = std::fs::create_dir_all(cache_path) {
+                    warn!(target: "signaturesidentifier", "could not create signatures cache dir: {:?}", err);
+                }
+                CachedSignatures::default()
+            };
+            return Ok(Self { cached, cached_path: Some(path) })
+        }
+        Ok(Self::default())
+    }
+
+    pub fn save(&self) {
+        if let Some(cached_path) = &self.cached_path {
+            if let Ok(file) = std::fs::File::create(&cached_path) {
+                if serde_json::to_writer(BufWriter::new(file), &self.cached).is_err() {
+                    warn!(target: "signaturesidentifier", "could not serialize SignaturesIdentifier");
+                }
+            } else {
+                warn!(target: "signaturesidentifier", "could not open {}", cached_path.to_string_lossy());
+            }
+        }
```

Comment:
wonder if it's time we switched all our caching from json to use sled and dump everything to a single embedded db file

Reply:
yes please :pray: